File: solver/solver.go
package solver

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

var (
	proxies    []string
	proxyIndex uint64
	proxyOnce  sync.Once
)

func loadProxies() {
	proxyOnce.Do(func() {
		f, err := os.Open("solving_proxies.txt")
		if err != nil {
			log.Fatalf("Failed to open proxies.txt: %v", err)
		}
		defer f.Close()

		scanner := bufio.NewScanner(f)
		for scanner.Scan() {
			line := strings.TrimSpace(scanner.Text())
			if line != "" {
				proxies = append(proxies, line)
			}
		}
		if len(proxies) == 0 {
			log.Fatal("proxies.txt is empty")
		}
		log.Printf("Loaded %d proxies", len(proxies))
	})
}

func nextProxy() string {
	idx := atomic.AddUint64(&proxyIndex, 1) - 1
	return proxies[idx%uint64(len(proxies))]
}

type TurnstileResponse struct {
	Success bool   `json:"success"`
	Token   string `json:"result"`
	Error   string `json:"error"`
}

func SolveTurnstile() (string, error) {
	loadProxies()
	proxy := nextProxy()

	payload := map[string]string{
		"type":     "turnstile",
		"site_key": "0x4AAAAAAA9oPHboisPr8cag",
		"url":      "https://graph.amctheatres.com/",
		"action":   "login",
		"proxy":    proxy,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("failed to marshal payload: %w", err)
	}

	req, err := http.NewRequest("POST", "http://38.22.104.125:3003/turnstile_solve", bytes.NewReader(body))
	if err != nil {
		return "", fmt.Errorf("failed to build request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-API-Key", "nsl_d93dfc0eebe91ad945bec902f57e4e4ec85fe6af018352d4")

	httpClient := &http.Client{Timeout: 30 * time.Second}
	resp, err := httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	var result TurnstileResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("failed to decode response: %w", err)
	}

	if !result.Success {
		return "", fmt.Errorf("solver error: %s", result.Error)
	}

	return result.Token, nil
}

File: rewrite_rust/src/solver.rs
use anyhow::{bail, Result};
use serde::{Deserialize, Serialize};
use std::time::Duration;

#[derive(Serialize)]
struct TurnstilePayload {
    r#type: String,
    site_key: String,
    url: String,
    action: String,
}

#[derive(Deserialize)]
struct TurnstileResponse {
    success: bool,
    token: Option<String>,
    error: Option<String>,
}

pub async fn solve_turnstile() -> Result<String> {
    let payload = TurnstilePayload {
        r#type: "turnstile".into(),
        site_key: "0x4AAAAAAA9oPHboisPr8cag".into(),
        url: "https://graph.amctheatres.com/".into(),
        action: "login".into(),
    };

    let client = wreq::Client::builder()
        .timeout(Duration::from_secs(30))
        .build()?;

    let resp = client
        .post("https://api.nslsolver.com/solve")
        .header("X-API-Key", "nsl_d93dfc0eebe91ad945bec902f57e4e4ec85fe6af018352d4")
        .header("Content-Type", "application/json")
        .json(&payload)
        .send()
        .await?;

    let result: TurnstileResponse = resp.json().await?;

    if !result.success {
        bail!(
            "solver error: {}",
            result.error.unwrap_or_else(|| "unknown".into())
        );
    }

    result
        .token
        .ok_or_else(|| anyhow::anyhow!("solver returned success but no token"))
}
